skimage/_shared/utils.py

```diff
@@ -20,6 +20,45 @@ class skimage_deprecation(Warning):
     pass
 
 
+class deprecate_arg:
+    """Decorator ensuring backward compatibility when argument names are
+    modified in a function definition.
+
+    Parameters
+    ----------
+    arg_mapping: dict
+        mapping between the function's old argument names and the new ones.
+    removed_version : str
+        The package version in which the deprecated argument will be removed.
+
+    """
+
+    def __init__(self, arg_mapping, removed_version=None):
+        self.arg_mapping = arg_mapping
+        self.warning_msg = ("'{old_arg}' is a deprecated argument name "
+                            "for `{func_name}`. ")
+        if removed_version is not None:
+            self.warning_msg += (f"It will be removed in version "
+                                 "{removed_version}. ")
+        self.warning_msg += "Use '{new_arg}' instead."
+
+    def __call__(self, func):
+        @functools.wraps(func)
+        def fixed_func(*args, **kwargs):
+            for old_arg, new_arg in self.arg_mapping.items():
+                if old_arg in kwargs:
+                    #  warn that the function interface has changed:
+                    warnings.warn(self.warning_msg.format(
+                        old_arg=old_arg, func_name=func.__name__,
+                        new_arg=new_arg), FutureWarning, stacklevel=2)
```

Comment:
Is this stack level correct?

Reply:
Good question! I made this simple test:

```python
from functools import wraps
import warnings


def my_decorator(f):
    @wraps(f)
    def wrapper(*args, **kwargs):
        warnings.warn("Oooooops", stacklevel=2)
        return f(*args, **kwargs)
    return wrapper


@my_decorator
def func():
    print("Here I am")


if __name__ == '__main__':
    func()
```
I saved this code snippet to a script `test.py` and got after execution
```
python test.py 
test.py:19: UserWarning: Oooooops
  func()
Here I am
```
So yes, the stack level is correct ;-)

Reply:
Thanks!

Reply:
> I think it is, as long as this decorator is the last one that was applied. Maybe that should go into the notes section unless we want to make this a parameter for the decorator.

Didn't notice that you already replied but I think it's till relevant.